src/Init/Data/Nat/Bitwise/Lemmas.lean

```diff
@@ -115,6 +115,35 @@ theorem testBit_add (x i n : Nat) : testBit x (i + n) = testBit (x / 2 ^ n) i :=
     rw [← Nat.add_assoc, testBit_add_one, ih (x / 2),
       Nat.pow_succ, Nat.div_div_eq_div_mul, Nat.mul_comm]
 
+theorem testBit_mul_two_pow (x i n : Nat) :
+    testBit (x * 2 ^ n) i = if n ≤ i then testBit x (i - n) else false := by
+  simp only [testBit, one_and_eq_mod_two, mod_two_bne_zero, Bool.if_false_right]
+  by_cases hni : n ≤ i
+  · simp only [hni, decide_true, Bool.true_and]
+    congr 2
+    let j := i - n
+    have hj : (x * 2 ^ n) >>> i = (x * 2 ^ n) >>> (j + n) :=  by simp only [j]; rw [Nat.sub_add_cancel]; omega
+    have hj' : x >>> (i - n) = x >>> j :=  by simp only [j]
+    rw [hj, hj', ← shiftLeft_eq, Nat.add_comm, shiftRight_add, shiftLeft_shiftRight]
+  · simp only [hni, decide_false, Bool.false_and, beq_eq_false_iff_ne, ne_eq]
+    simp only [Nat.not_le] at hni
+    rw [← shiftLeft_eq]
+    let k := n - i
+    have hk : x <<< n >>> i = x <<< (k + i) >>> i := by simp only [k]; rw [Nat.sub_add_cancel]; omega
+    rw [hk, Nat.shiftLeft_add, Nat.shiftLeft_shiftRight, shiftLeft_eq]
+    have hk' : 0 < k := by omega
+    have hx : 2 * (x * 2 ^ k / 2) = x * 2 ^ k := by
+      rw [Nat.mul_comm, Nat.div_mul_cancel]
+      suffices hs : 2 * (x * 2 ^ (k - 1)) = x * 2 ^ k by
+        exact ⟨_, hs.symm⟩
+      let j := k - 1
+      have hj : 2 ^ (k - 1) = 2 ^ j := by simp only [j]
+      have hj' : 2 ^ k = 2 ^ (j + 1) := by simp only [j]; rw [Nat.sub_add_cancel]; omega
+      rw [hj, hj']
+      simp only [Nat.pow_add, Nat.pow_one, j, k]
+      rw [Nat.mul_comm, Nat.mul_assoc]
+    omega
```

Comment:
Could you separate out the two branches of the proof into their own lemmas: `testBit_mul_two_pow_of_le` that takes in `n <= i` as hypothesis and `testBit_mul_two_pow_of_lt` that takes in `i < n` (feel free to come up with other names, especially if they convey the flip in variable order more clearly). Each of these will look a bit more manageable, and then this proof becomes a trivial consequence.